xtrack/beam_elements/elements.py

```diff
@@ -336,6 +336,81 @@ def get_backtrack_element(self, _context=None, _buffer=None, _offset=None):
                               _context=_context, _buffer=_buffer, _offset=_offset)
 
 
+class FastQuadrupole(BeamElement):
+    """An optimised version of a quadrupole with zero knl[0], ksl, hxl, hyl, and length.
+    Parameters:
+
+            - knl [m^-n, array]: Normalized integrated strength of the normal components.
+
+    """
+
+    _xofields={
+        'knl': xo.Float64[2],
+    }
+
+    _extra_c_sources = [
+        xp.general._pkg_root.joinpath('random_number_generator/rng_src/base_rng.h'),
+        xp.general._pkg_root.joinpath('random_number_generator/rng_src/local_particle_rng.h'),
+        _pkg_root.joinpath('beam_elements/elements_src/fastquadrupole.h')]
+
+    def __init__(self, knl=None, **kwargs):
+
+        if '_xobject' in kwargs.keys() and kwargs['_xobject'] is not None:
+            self.xoinitialize(**kwargs)
+            return
+
+        if len(knl) != 2:
+            raise ValueError("For a quadrupole, len(knl) must be 2.")
+
+        kwargs["knl"] = knl
+        self.xoinitialize(**kwargs)
+
+    def get_backtrack_element(self, _context=None, _buffer=None, _offset=None):
+        ctx2np = self._buffer.context.nparray_from_context_array
+        return self.__class__(knl=-ctx2np(self.knl), _context=_context,
+                              _buffer=_buffer, _offset=_offset)
+
+
+class FastDipole(BeamElement):
+    """An optimised version of a dipole with zero ksl and hyl. Parameters:
+
+            - knl [m^-n, array]: Normalized integrated strength of the normal components.
+            - hxl [rad]: Rotation angle of the reference trajectory in the horizontal plane.
+            - length [m]: Length of the originating thick multipole.
+
+    """
+
+    _xofields={
+        'knl': xo.Float64[1],
+        'hxl': xo.Float64,
```

Comment:
We can assume hxl = knl[0]. Maybe it gives us a visible speedup?

Reply:
I ran it a few times and it seems that it is consistently faster, but only by like 0.3% on average on the simulation you gave me. There is a difference in the number of elements that can be converted too, and it's similarly insignificant.